### PR TITLE
feature - login viewmodel

### PR DESCRIPTION
**Pull request by at-duytran**

### CODE REVIEW

**Comment by at-daiho** [comment]
:+1: There are no linter violations.

File: MyApp/Define/Strings.swift [hunk 1]
```diff
@@ -18,5 +18,6 @@ extension App {
     struct String {
         static let error = "ERROR".localized()
         static let ok = "OK".localized()
+        static let newItemCollectionCell = "NewItemCollectionCell".localized()
```
Comment (at-truongnguyen) [inline MyApp/Define/Strings.swift]: what is this use for?

File: MyApp/View/Controllers/Login/LoginViewController.swift [hunk 2]
```diff
@@ -52,9 +60,29 @@ final class LoginViewController: UIViewController {
...
+    enum GetLoginResult {
+        case success
+        case failure
+    }
+    typealias GetLoginCompletion = (GetLoginResult) -> Void
+
+    @IBAction private func loginButtonTouchUpInside(_ sender: Any) {
+        guard let username = userNameTextField.text else { return }
+        guard let password = passWordTextField.text else { return }
+        viewModel.checkLoginInformation(username: username, password: password)
```
Comment (at-truongnguyen) [inline MyApp/View/Controllers/Login/LoginViewController.swift]: Is this validate? If that i think we need validate Success or Error. Please take a look at `LoginviewController` in Demo project.

File: MyApp/ViewModel/Login/LoginViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import MVVM
+
+final class LoginViewModel: MVVM.ViewModel {
+    var userName = ""
+    var passWord = ""
+    var loginStatus = ""
+
+    enum GetLoginResult {
+        case success
+        case failure
```
Comment (at-truongnguyen) [inline MyApp/ViewModel/Login/LoginViewModel.swift]: if error, I think we should include error in it.

File: MyApp/ViewModel/Login/LoginViewModel.swift [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    typealias GetLoginCompletion = (GetLoginResult) -> Void
+
+    func checkLoginInformation(username: String, password: String) {
+        userName = username
+        passWord = password
+    }
+
+    func getLoginStatus(_ completion: @escaping GetLoginCompletion) {
+        if userName.isEmpty || passWord.isEmpty {
+            loginStatus = "[!] Username & Password must be filled"
```
Comment (at-truongnguyen) [inline MyApp/ViewModel/Login/LoginViewModel.swift]: Define this Text please

File: MyApp/ViewModel/Login/LoginViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+            completion(.failure)
+        } else {
+            if userName != LoginModel.shared.userAccount.userName {
+                loginStatus = "[!] Username is not available"
```
Comment (at-truongnguyen) [inline MyApp/ViewModel/Login/LoginViewModel.swift]: define also

File: MyApp/ViewModel/Login/LoginViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+                completion(.failure)
+            } else if userName == LoginModel.shared.userAccount.userName && passWord != LoginModel.shared.userAccount.passWord {
```
Comment (at-truongnguyen) [inline MyApp/ViewModel/Login/LoginViewModel.swift]: This one will be delete after apply api, right?

File: MyApp/View/Controllers/Login/LoginViewController.swift [hunk 7]
```diff
@@ -52,9 +55,43 @@ final class LoginViewController: UIViewController {
...
+    func textField(for field: LoginViewModel.Field) -> UITextField {
+        switch field {
+        case .userName: return userNameTextField
+        case .passWord: return passWordTextField
+        }
+    }
+
+    enum GetLoginResult {
+        case success
+        case failure
```
Comment (at-truongnguyen) [inline MyApp/View/Controllers/Login/LoginViewController.swift]: why we don't have error here?

File: MyApp/View/Controllers/Login/LoginViewController.swift [hunk 8]
```diff
@@ -52,9 +55,43 @@ final class LoginViewController: UIViewController {
...
+        switch viewModel.validate() {
+        case .success:
+            viewModel.login { [weak self] (result) in
+                guard let this = self else { return }
+                switch result {
+                case .success:
+                    let vc = HomeViewController()
+                    this.navigationController?.pushViewController(vc, animated: true)
+                case .failure:
+                    break
```
Comment (at-truongnguyen) [inline MyApp/View/Controllers/Login/LoginViewController.swift]: if `failure` please show error.

File: MyApp/View/Controllers/Login/LoginViewController.swift [hunk 9]
```diff
@@ -52,9 +55,43 @@ final class LoginViewController: UIViewController {
...
+                }
+            }
+        case .failure(let field, let msg):
+            alert(msg: msg, buttons: [App.Strings.ok], handler: { [weak self] _ in
+                guard let this = self else { return }
+                let fld = this.textField(for: field)
+                fld.becomeFirstResponder()
+            })
+            viewDidUpdated()
```
Comment (at-truongnguyen) [inline MyApp/View/Controllers/Login/LoginViewController.swift]: Please notice that this `viewDidUpdated ` for writing UnitTest

File: MyApp/ViewModel/Login/LoginViewModel.swift [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+        guard passWord.len >= 7 else {
+            return .failure(field: .passWord, msg: "'\(Field.passWord)' \(App.Strings.tooShort)")
+        }
+        guard userName == LoginModel.shared.userAccount.userName else {
+            return .failure(field: .userName, msg: "\(App.Strings.wrongUsername) '\(userName)'")
+        }
+        guard passWord == LoginModel.shared.userAccount.passWord else {
+            return .failure(field: .passWord, msg: "\(App.Strings.wrongPassword)")
+        }
+        return .success
```
Comment (at-truongnguyen) [inline MyApp/ViewModel/Login/LoginViewModel.swift]: Please follow API document then validate this 

Email: The email suffix format must be @asiantech.vn
Password: Must contain 6 - 255 characters

File: MyApp/ViewModel/Login/LoginViewModel.swift [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        var description: String {
+            switch self {
+            case .success: return "\(App.Strings.success)"
+            case .failure(_, let msg):
+                return "\(App.Strings.failure): " + msg
+            }
+        }
+    }
+
+    var userName = ""
```
Comment (at-truongnguyen) [inline MyApp/ViewModel/Login/LoginViewModel.swift]: We login by email, not by username

File: MyApp/ViewModel/Login/LoginViewModel.swift [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+    func validate() -> Validation {
+        let checkEmailFormat = isValidEmail(eMail: eMail)
+        guard checkEmailFormat == true else {
+            return .failure(field: .eMail, msg: App.Strings.wrongEmailFormat)
+        }
+        guard passWord.len >= 6 && passWord.len <= 255 else {
+            return .failure(field: .passWord, msg: App.Strings.wrongPasswordLength)
+        }
+        guard eMail == LoginModel.shared.userAccount.eMail else {
```
Comment (at-truongnguyen) [inline MyApp/ViewModel/Login/LoginViewModel.swift]: why we validate this conditional?

File: MyApp/ViewModel/Login/LoginViewModel.swift [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+            return .failure(field: .eMail, msg: App.Strings.wrongEmail)
+        }
+        guard eMail == LoginModel.shared.userAccount.eMail && passWord == LoginModel.shared.userAccount.passWord else {
```
Comment (at-truongnguyen) [inline MyApp/ViewModel/Login/LoginViewModel.swift]: why we validate this conditional?

File: MyApp/Model/API/API.Me.swift [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+                    return
+                }
+
+                DispatchQueue.main.async {
+                    completion(.success(me))
+                }
+            case .failure(let error):
+                completion(.failure(error))
+            }
+        }
```
Comment (at-truongnguyen) [inline MyApp/Model/API/API.Me.swift]: Response when login api `api/v1/user/login`
{
  "data": {
    "userId": 1,
    "tokenType": "Bearer",
    "id": 2,
    "accessToken": "tHpG1AwyO0M3fioML58KyXME/jwgGdqowRIl6jyb/68=",
    "expireAt": "2017-08-02T07:22:28.735Z"
  }
}

**Pull request merged**